File: backend/api/src/shared/middleware/authenticate.py
"""JWT authentication dependency for FastAPI.

Ported from Node.js createAuthenticateHook — same contract, FastAPI Depends pattern.

Usage in routes:
    from fastapi import Depends
    from src.shared.middleware.authenticate import get_current_user

    @router.get("/protected")
    async def protected(user=Depends(get_current_user)):
        ...
"""

from typing import Optional

from fastapi import Header

from src.modules.identity.services.auth_service import AuthService
from src.shared.domain.errors import TokenExpiredError, TokenInvalidError, UnauthorizedError


def authenticate_dependency(auth_service: AuthService):
    """Create a dependency function that verifies JWT tokens.

    Returns a callable suitable for use with FastAPI Depends().
    The callable extracts the Bearer token from the Authorization header,
    verifies it, and returns the decoded user dict {id, email, role}.
    """

    def _authenticate(authorization: Optional[str] = Header(None)) -> dict:
        if not authorization or not authorization.startswith("Bearer "):
            raise UnauthorizedError("Missing or malformed Authorization header")

        token = authorization[7:]

        # verify_access_token raises TokenExpiredError or TokenInvalidError
        payload = auth_service.verify_access_token(token)

        return {
            "id": payload["sub"],
            "email": payload["email"],
            "role": payload["role"],
        }

    return _authenticate

File: backend/services/_archived/identity/src/middleware/authenticate.ts
import type { FastifyRequest, FastifyReply } from 'fastify';
import type { AuthService } from '../services/AuthService.js';
import type { ILogger } from '../logger.js';
import {
  TokenExpiredError,
  TokenInvalidError,
  UnauthorizedError,
} from '../domain/errors.js';
import type { UserRole } from '../domain/User.js';

export interface AuthenticatedUser {
  id: string;
  email: string;
  role: UserRole;
}

// Augment Fastify's Request type
declare module 'fastify' {
  interface FastifyRequest {
    user?: AuthenticatedUser;
    traceId?: string;
  }
}

export function createAuthenticateHook(authService: AuthService, logger: ILogger) {
  return async function authenticate(
    request: FastifyRequest,
    reply: FastifyReply,
  ): Promise<void> {
    const start = Date.now();
    const traceId = (request.headers['x-trace-id'] as string | undefined) ?? request.traceId ?? 'unknown';
    const childLogger = logger.child({ trace_id: traceId });

    const authHeader = request.headers.authorization;
    if (!authHeader || !authHeader.startsWith('Bearer ')) {
      childLogger.warn('Missing or malformed Authorization header', {
        duration_ms: Date.now() - start,
      });
      await reply.status(401).send({
        success: false,
        error: { code: 'UNAUTHORIZED', message: 'Missing or malformed Authorization header' },
        meta: { traceId },
      });
      return;
    }

    const token = authHeader.slice(7);

    try {
      const payload = authService.verifyAccessToken(token);
      request.user = {
        id: payload.sub,
        email: payload.email,
        role: payload.role as UserRole,
      };

      childLogger.debug('JWT verified', {
        userId: payload.sub,
        duration_ms: Date.now() - start,
      });
    } catch (err) {
      const duration_ms = Date.now() - start;

      if (err instanceof TokenExpiredError) {
        childLogger.warn('JWT expired', { duration_ms });
        await reply.status(401).send({
          success: false,
          error: { code: 'TOKEN_EXPIRED', message: 'Token has expired' },
          meta: { traceId },
        });
        return;
      }

      if (err instanceof TokenInvalidError || err instanceof UnauthorizedError) {
        childLogger.warn('Invalid JWT', { duration_ms });
        await reply.status(401).send({
          success: false,
          error: { code: 'TOKEN_INVALID', message: 'Invalid token' },
          meta: { traceId },
        });
        return;
      }

      childLogger.error('Unexpected error during authentication', {
        duration_ms,
        error: err instanceof Error ? err : new Error(String(err)),
      });
      await reply.status(401).send({
        success: false,
        error: { code: 'UNAUTHORIZED', message: 'Authentication failed' },
        meta: { traceId },
      });
    }
  };
}
